core-to-plc/src/Language/Plutus/CoreToPLC/Compiler/Builtins.hs

```diff
@@ -73,19 +73,42 @@ Haskell. But there is a problem: they use either primitive or Scott-encoded vers
 whereas when we compile them from Haskell they will end up as abstract types, and so the types
 won't line up at the call site.
 
+That is, we generate something like this:
+(/\ (Int :: *) .
+  (\ addInteger : Int -> Int -> Int .
+      <use addInteger>
+  )
+  (\ x,y : Int . <builtin addInteger> x y) -- Uh oh, type error, can't show that Int = <builtin int>!
+)
+{<builtin int>}
+
 We handle this in two different ways:
-- For the types like Bool and Unit which are really algebraic type, and which have constructors etc.
+- For the types like Bool and Unit which are really algebraic types, and which have constructors etc.
 that we care about elsewhere, we insert adaptor code into the definition of the builtin (see note [Mapping builtins]).
 - For types like Int and Bytestring which don't have visible constructors, we can treat them as completely opaque,
 and we use a type alias. (Actually we fake the alias until we have aliases in PIR)
 
-Why do we need an alias rather than just a normal type def? The problem is that the type we want to present to the
-user for `addInteger` is `Int -> Int -> Int`, but in the *body* of the definition the argument types need to be
-*PLC* ints otherwise the builtin won't be happy. So we need to break the abstraction in the body, hence the alias.
+Here's how that looks for a primitive that takes Ints and returns a Boolean, assuming we have bound Int and Bool
+already as an alias and an abstract type respectively:
+(\ equalsInteger : Int -> Int -> Bool .
+  <use equalsInteger>
+)
+(\ x, y : Int . -- No need to do anything to the arguments, since we're using a transparent alias for Int
+  (<builtin equalsInteger> x y) {Bool} True False -- Immediately match the builtin bool into an abstract Bool
+)
 
 We *could* do something like the interleaved definitions that we do for datatypes in PIR. Morally this is perhaps the
-right thing to do: we should thing of Int and its builtins as a "module" that goes together and where all the definitions
-have access to the internals of the other definitions. However, for the moment this seems more trouble than it's worht.
+right thing to do: we should think of Int and its builtins as a "module" that goes together and where all the definitions
+have access to the internals of the other definitions. However, for the moment this seems more trouble than it's worth.
+
+For an example of how that would look:
+(/\ (Int :: *) .
+  (\ addInteger : Int -> Int -> Int . -- Type signature is fine inside the abstraction
+      <use addInteger>
+  )
+)
+{<builtin int>}
+(\ x,y : <builtin int> . <builtin addInteger> x y) -- No type error any more, abstraction is gone
```

Comment:
This very much reminds me of how data types and their constructors/destructors are handled. Is there any code that is/can be shared?

Reply:
Yes, as I say in the previous paragraph, it's just like what we do for datatypes. I've been thinking about this a little: if we could add some kind of construct to PIR that did this, then we could indeed share the PIR codegen between them. In a sense the "abstract datatype" pattern is an instance of the "abstract module" pattern. So I think something in this direction is promising.